### PR TITLE
Edge switchover extension work

### PR DESCRIPTION
**Pull request by Artonus**
Added:
* Ability to move a single instance from one Middleware to another within a single organization

### CODE REVIEW

File: src/TaskPlanner/Contracts/Requests/PerformSwitchoverRequest.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Middleware.TaskPlanner.Contracts.Requests;
+
+public record PerformSwitchoverRequest
+{
+    public Guid ActionPlanId { get; init; }
+    public Guid InstanceId { get; init; }
+    public string Destination { get; init; }
```
Comment (adrianLIrobotics) [inline src/TaskPlanner/Contracts/Requests/PerformSwitchoverRequest.cs:7]: Is Destionation attribute the name of the cloud or edge?

File: src/TaskPlanner/Contracts/Requests/PerformSwitchoverRequest.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Middleware.TaskPlanner.Contracts.Requests;
+
+public record PerformSwitchoverRequest
+{
+    public Guid ActionPlanId { get; init; }
+    public Guid InstanceId { get; init; }
```
Comment (adrianLIrobotics) [inline src/TaskPlanner/Contracts/Requests/PerformSwitchoverRequest.cs:6]: We can only switch over one instance at a time? We cannot have a List<InstanceId> or a Dicctionary of InstanceId and DestionationName

File: src/Common/MessageContracts/SwitchoverDeployInstance.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+﻿namespace Middleware.Common.MessageContracts;
+
+public record SwitchoverDeployInstance : Message
```
Comment (adrianLIrobotics) [inline src/Common/MessageContracts/SwitchoverDeployInstance.cs:3]: Switch over request should be validated by middleware. By this I mean that it should take policies into consideration. A switch over to a cloud with a policy action for always using edge should not active switch over

File: src/DataAccess/Repositories/RedisRepository.cs [hunk 4]
```diff
@@ -67,6 +67,12 @@ public async Task<TModel> AddAsync(TModel model, Func<Guid> guidProvider)
             {
                 return null!;
             }
+
+            if (IsWritableToGraph)
```
Comment (adrianLIrobotics) [inline src/DataAccess/Repositories/RedisRepository.cs:71]: Are this the fixes for the errors we encouterted when trying to create OWNS relationship between robot and actionPlan?

File: src/Orchestrator/Deployment/DeploymentService.cs [hunk 5]
```diff
@@ -316,35 +320,96 @@ public async Task<bool> DeletePlanAsync(ActionPlanModel actionPlan)
...
+        var instance = actionPlan.ActionSequence
+            .SelectMany(x => x.Services!).FirstOrDefault(i => i.ServiceInstanceId == instanceId);
+        if (instance is null)
+            return;
+
+        BaseModel thisLocation = _mwConfig.Value.InstanceType == LocationType.Cloud.ToString()
+            ? await _redisInterfaceClient.GetCloudByNameAsync(_mwConfig.Value.InstanceName)
+            : await _redisInterfaceClient.GetEdgeByNameAsync(_mwConfig.Value.InstanceName);
+
+        await _redisInterfaceClient.DeleteRelationAsync(instance, thisLocation, "LOCATED_AT");
```
Comment (adrianLIrobotics) [inline src/Orchestrator/Deployment/DeploymentService.cs:340]: Perfect!

File: src/TaskPlanner/Contracts/Requests/PerformSwitchoverRequest.cs (hunk 1 above)
Comment (Artonus) [reply src/TaskPlanner/Contracts/Requests/PerformSwitchoverRequest.cs:7]: yes

File: src/Common/MessageContracts/SwitchoverDeleteInstance.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
+﻿namespace Middleware.Common.MessageContracts;
+
+public record SwitchoverDeleteInstance : Message
```
[the remark this answers is not in the document]
Comment (Artonus) [reply src/Common/MessageContracts/SwitchoverDeleteInstance.cs:3]: We can think about it when we will continue the work on the semantic planner

**Review by adrianLIrobotics — APPROVED** [review]
Most important validation of a switch over request may be needed to comply with active policies of the system

File: src/DataAccess/Repositories/RedisRepository.cs (hunk 4 above)
Comment (Artonus) [reply src/DataAccess/Repositories/RedisRepository.cs:71]: yes, they are

File: src/Common/MessageContracts/SwitchoverDeployInstance.cs (hunk 3 above)
Comment (Artonus) [reply src/Common/MessageContracts/SwitchoverDeployInstance.cs:3]: I am fully onboard with the idea. But I believe that first, we should revise how the policies are supposed to work in a separate ticket. When we will have an idea we can start plugging them into the parts of the system

File: src/TaskPlanner/Contracts/Requests/PerformSwitchoverRequest.cs (hunk 2 above)
Comment (Artonus) [reply src/TaskPlanner/Contracts/Requests/PerformSwitchoverRequest.cs:6]: Moving more than one instance would be beneficial for us. But I'm afraid I have to disagree with passing a list of deployed instance identifiers. I want to suggest the middle ground. We will pass the id of action, allowing us to move all the instances deployed within a single action. What do you think?

Comment (Artonus) [reply src/TaskPlanner/Contracts/Requests/PerformSwitchoverRequest.cs]: Changed

**Pull request merged**